File: app/src/main/java/com/cqebd/student/http/NetClient.java
package com.cqebd.student.http;

import android.text.TextUtils;
import android.util.SparseArray;

import com.cqebd.student.net.gateway.GatewayInterceptor;
import com.orhanobut.logger.Logger;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;

import gorden.util.XLog;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * document
 * Created by Gordn on 2017/2/28.
 */
public class NetClient {

    private static SparseArray<List<Call>> attachCalls = new SparseArray<>();

    private static final String BASE_URL = "http://service.ex.cqebd.cn/";
    private static final String TAG = "http_log";

    private static OkHttpClient okHttpClient = new OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .readTimeout(300, TimeUnit.SECONDS)
            .addInterceptor(new GatewayInterceptor("23393048", "d0c983467d8ced6568e844c0b0a233ae"))
            .addInterceptor(new HttpLoggingInterceptor(message -> {
                if (TextUtils.isEmpty(message)) return;
                if (message.startsWith("{") || message.startsWith("[")) {
                    XLog.e(TAG, message);
                }
            }).setLevel(HttpLoggingInterceptor.Level.BODY))
            .build();

    private static Retrofit retrofit = new Retrofit.Builder().baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create()).client(okHttpClient).build();

    private NetClient() {
    }

    public static OkHttpClient getOkHttpClient() {
        return okHttpClient;
    }

    public static <T> T createApi(Class<T> clazz) {
        return retrofit.create(clazz);
    }

    public static void netErr(String errMsg) {
        if (!TextUtils.isEmpty(errMsg)) {
            Logger.d(errMsg);
        }
    }

    public static void cancel(int key) {
        List<Call> calls = attachCalls.get(key);
        if (calls != null) {
            for (Call call : calls) {
                call.cancel();
                XLog.e("取消一个请求");
            }
            attachCalls.remove(key);
            XLog.e("移除key = " + key + "  的请求");
        }
    }

    public static void request(int key, Call call, Callback callback) {
        if (attachCalls.get(key) != null) {
            XLog.e("已有key 里添加请求");
            attachCalls.get(key).add(call);
        } else {
            List<Call> calls = new ArrayList<>();
            calls.add(call);
            XLog.e("生成key的请求");
            attachCalls.append(key, calls);
        }
        XLog.e("当前存放的请求:" + Arrays.toString(attachCalls.get(key).toArray()));
        call.enqueue(callback);
    }
}

File: app/src/main/java/com/cqebd/student/net/NetClient.kt
package com.cqebd.student.net

import com.cqebd.student.app.App
import com.cqebd.student.net.adapter.LiveDataCallAdapterFactory
import com.cqebd.student.net.api.ApiService
import com.cqebd.student.net.api.VideoService
import com.cqebd.student.net.api.WorkService
import com.cqebd.student.net.converter.ApiConverterFactory
import com.cqebd.student.net.gateway.GatewayInterceptor
import com.orhanobut.logger.Logger
import com.readystatesoftware.chuck.ChuckInterceptor
import gorden.lib.anko.static.logWarn
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.util.concurrent.TimeUnit

/**
 * 描述
 * Created by gorden on 2018/3/1.
 */
object NetClient {
    private const val TIME_OUT = 15L

    private val workClient by lazy {
        OkHttpClient.Builder()
                .connectTimeout(TIME_OUT, TimeUnit.SECONDS)
                .writeTimeout(TIME_OUT, TimeUnit.SECONDS)
                .readTimeout(TIME_OUT, TimeUnit.SECONDS)
                .addInterceptor(ChuckInterceptor(App.mContext))
                .addInterceptor(GatewayInterceptor("23393048", "d0c983467d8ced6568e844c0b0a233ae"))
                .addInterceptor(HttpLoggingInterceptor { message ->
                    logWarn(message, "http_log")
                }.setLevel(HttpLoggingInterceptor.Level.BODY))
                .build()
    }

    private val videoClient by lazy {
        OkHttpClient.Builder()
                .connectTimeout(TIME_OUT, TimeUnit.SECONDS)
                .writeTimeout(TIME_OUT, TimeUnit.SECONDS)
                .readTimeout(TIME_OUT, TimeUnit.SECONDS)
                .addInterceptor(ChuckInterceptor(App.mContext))
                .addInterceptor(GatewayInterceptor("23776862", "b5ffc0cc02a74953ea9091338117feda"))
                .addInterceptor(HttpLoggingInterceptor { message ->
                    logWarn(message, "http_log")
                }.setLevel(HttpLoggingInterceptor.Level.BODY))
                .build()
    }

    private var workService: WorkService? = null
    fun workService(): WorkService {
        if (workService == null) {
            synchronized(NetClient::class.java) {
                if (workService == null) {
                    workService = Retrofit.Builder().baseUrl("http://service.ex.cqebd.cn/")
                            .addCallAdapterFactory(LiveDataCallAdapterFactory())
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                            .addConverterFactory(ApiConverterFactory.create())
                            .client(workClient)
                            .build().create(WorkService::class.java)
                }
            }
        }
        return workService!!
    }

    private var videoService: VideoService? = null
    fun videoService(): VideoService {
        if (videoService == null) {
            synchronized(NetClient::class.java) {
                if (videoService == null) {
                    videoService = Retrofit.Builder().baseUrl("http://service.onlin.cqebd.cn/")
                            .addCallAdapterFactory(LiveDataCallAdapterFactory())
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                            .addConverterFactory(ApiConverterFactory.create())
                            .client(videoClient)
                            .build().create(VideoService::class.java)
                }
            }
        }
        return videoService!!
    }

    private var apiService: ApiService? = null
    fun apiService(): ApiService {
        if (apiService == null) {
            synchronized(NetClient::class.java) {
                if (apiService == null) {
                    apiService = Retrofit.Builder().baseUrl("http://service.student.cqebd.cn/")
                            .addCallAdapterFactory(LiveDataCallAdapterFactory())
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                            .addConverterFactory(ApiConverterFactory.create())
                            .build().create(ApiService::class.java)
                }
            }
        }
        return apiService!!
    }
}